Compound integration with Omen Markets. Archived in favor of https://github.com/protofire/omen-exchange/pull/1570

pRoy24 opened the pull request:
Fixed the previous round of comments.
Added badge as defined in the mockup.
Refactored to make it work with ETH as the base and cETH as the underlying collateral.
Tested on Rinkeby with ETH -> cETH and dai -> cDAI pairs.

Please note that you need to run yarn install since this feature uses Big.js version 6.0.0

pimato commented:
Hey @pRoy24 and thanks for submitting your Pull Request. There are several improvements we can make to improve the user experience but overall really great work!

pimato commented:
I think it is important to add an indicator to the market details view that this market is actually earning interest via compound. This is my first attempt for it:

<img width="712" alt="Details" src="https://user-images.githubusercontent.com/5337809/103639863-5448cd00-4f4f-11eb-9ae1-161601b445fb.png">

**Figma**: 
[Design Specification](https://www.figma.com/file/R056tSYMjvfwaFidEBItYy/Widgets-V2?node-id=1209%3A141)

Mi-Lan commented on app/src/components/market/common/add_compound_service/index.tsx:
```diff
@@ -0,0 +1,118 @@
+  -webkit-box-align: center;
+`
+
+const ServiceTextWrapper = styled.div`
+  width: 90%;
+`
+
+const ServiceCheckWrapper = styled.div`
+  width: 10%;
+`
```
wondering if we could do without these two classes?

pRoy24 replied:
I think this is needed to enforce the width of the respective containers

Mi-Lan commented on app/src/components/market/common/add_compound_service/index.tsx:
```diff
@@ -0,0 +1,118 @@
+    serviceChecked = <span />
+  }
+  return (
+    <Wrapper>
+      <Title>Recommended Service</Title>
+      <DescriptionWrapper>
+        <Description>
+          <ServiceWrapper>
+            <ServiceIconWrapper>
+              <CompoundIcon />
+            </ServiceIconWrapper>
+            <ServiceTokenDetails>
+              <ServiceTextWrapper>
+                <div>Compound</div>
+                <div>
+                  Convert {currentToken} to c{currentToken} to earn {compoundInterestRate}% interest
+                </div>
+              </ServiceTextWrapper>
+              <ServiceCheckWrapper onClick={toggleServiceCheck}>
+                <CheckService>{serviceChecked}</CheckService>
+              </ServiceCheckWrapper>
+            </ServiceTokenDetails>
+          </ServiceWrapper>
+        </Description>
+      </DescriptionWrapper>
+    </Wrapper>
```
Maybe we can check what @KadenZipfel  thinks since he is more senior developer but i believe this whole section has too many unnecessary classes and could be reduced...

Mi-Lan commented on app/src/components/market/common/add_compound_service/index.tsx:
```diff
@@ -0,0 +1,118 @@
+                  Convert {currentToken} to c{currentToken} to earn {compoundInterestRate}% interest
+                </div>
+              </ServiceTextWrapper>
+              <ServiceCheckWrapper onClick={toggleServiceCheck}>
+                <CheckService>{serviceChecked}</CheckService>
+              </ServiceCheckWrapper>
```
![Screenshot 2021-01-05 at 13 08 55](https://user-images.githubusercontent.com/33226956/103644795-3bdcb080-4f57-11eb-9179-09e3fc082244.png)
Seems that this button is not center aligned. Also when it's clicked it should invert colors but now the way it is now checkmark just disappears...

Mi-Lan commented on app/src/components/market/sections/market_buy/market_buy.tsx:
```diff
@@ -256,14 +374,16 @@ const MarketBuyWrapper: React.FC<Props> = (props: Props) => {
                   setAmount(new BigNumber(0))
+                  setDisplayAmountToFund(new BigNumber(0))
                 }
               }}
```
![Screenshot 2021-01-05 at 12 45 49](https://user-images.githubusercontent.com/33226956/103645383-303db980-4f58-11eb-86c2-badaf52c8268.png)
I seem to be getting this error at random time? What is that related to?

pRoy24 replied:
This shoudl be resolved now

Mi-Lan reviewed (changes requested):
Hey men good job navigating our codebase seems that you have gotten the hang of it pretty fast! Also i'm having issue creating market? Do you know whats the problem here is the video...But regular markets work good i can create them...

https://user-images.githubusercontent.com/33226956/103646307-c1f9f680-4f59-11eb-8dcd-fbec8545fd8b.mov

KadenZipfel commented on app/src/components/market/sections/market_sell/market_sell.tsx:
```diff
@@ -38,6 +49,34 @@ const StyledButtonContainer = styled(ButtonContainer)`
+  display: inline-flex;
+  padding-left: 14px;
+  padding-top: 14px;
+  color: #37474f;
```
Color from `themes.js` here too

KadenZipfel commented on app/src/services/compound_service.ts:
```diff
@@ -0,0 +1,109 @@
+
+  calculateSupplyRateAPY = async (): Promise<number> => {
+    const supplyRate: number = await this.contract.supplyRatePerBlock()
+    const ethMantissa = 1e18
```
Should the mantissa be different for different token decimal amounts?

pRoy24 replied:
yeah I was a bit confused about this too since the docs have an example for ETH but the 1e18 mantissa seems to give the correct APY for all tokens regardless of the decimals so just changed the name of the constant to make it more generic but left the value as is.

KadenZipfel reviewed (changes requested):
This is quite a big change, and overall it's pretty well done. Some additional notes:

- `batch transaction execution` fails at market creation if compound service is enabled.
- `list_view` displays the cToken as the currency but elsewhere it's abstracted away.
    - ![](https://firebasestorage.googleapis.com/v0/b/firescript-577a2.appspot.com/o/imgs%2Fapp%2Fkaden-notes%2FQ5SqZG9So9.png?alt=media&token=e5c786c4-7c94-47dc-9ef8-5edb1346b9c6)
    - I'm assuming it should also be abstracted away here as well.
- Error depositing DAI in DAI market with compound service enabled:
    - ![](https://firebasestorage.googleapis.com/v0/b/firescript-577a2.appspot.com/o/imgs%2Fapp%2Fkaden-notes%2FyPzPnkB4wT.png?alt=media&token=75d39d0e-3e21-427f-84f5-11df6c3c70b4)
    - cDai deposit seems to work fine though
- Base token balance is sometimes incorrectly set to 0
    - ![](https://firebasestorage.googleapis.com/v0/b/firescript-577a2.appspot.com/o/imgs%2Fapp%2Fkaden-notes%2FRab6i_g2tM.png?alt=media&token=801760ca-d6b7-430f-bc82-ad5f7a548a9f)
- Outcome probabilities sometimes get messed up in the buy view
    - ![](https://firebasestorage.googleapis.com/v0/b/firescript-577a2.appspot.com/o/imgs%2Fapp%2Fkaden-notes%2FaJ8J7BH4Kc.png?alt=media&token=35220e97-e615-4da9-acb2-f0e3ad9ffad7)
- It seems like there may be some reused logic throughout components, wherever possible, we should make reusable functions in `tools.ts`

pimato reviewed (changes requested):
on funding page loaded, app freezes

pRoy24 commented on app/src/services/compound_abi.ts:
```diff
@@ -0,0 +1,15 @@
+/* eslint no-use-before-define: 0 */  // --> OFF
```
Moved the file to abi directory, I think prettifying makes the file larger and since we are not planning to edit the ABIs anyways, maybe just leave it as is and replace it with newer ABIs as Compound release new upgrades.

pimato commented:
Just got this error:
<img width="2241" alt="Bildschirmfoto 2021-01-11 um 15 02 47" src="https://user-images.githubusercontent.com/5337809/104191731-2adfe380-541e-11eb-8ba5-6dced21df5c4.png">

For this market:
<img width="613" alt="Bildschirmfoto 2021-01-11 um 15 04 14" src="https://user-images.githubusercontent.com/5337809/104191809-4a770c00-541e-11eb-81ad-a5cdf368eb6f.png">

pimato reviewed (changes requested):
see comment.

hexyls commented on app/src/services/cpk.ts:
```diff
@@ -280,6 +347,40 @@ class CPKService {
+            compoundTokenDetails.symbol,
+            marketData.funding.toString(),
+          )
+          // Approve cToken for the cpk contract
```
I had trouble creating markets with DAI, I think we might need to add an extra step here: Transfer the underlying token from the user to the CPK before we mint the cToken.

pRoy24 replied:
Yes, This was the case during an interim commit when I incorrectly resolved a merge conflict after rebasing to master. Should be fixed now :)

hexyls commented on app/src/components/market/common/add_compound_service/index.tsx:
```diff
@@ -0,0 +1,127 @@
+  currentToken: string
+}
+
+export const AddCompoundService: React.FC<AddCompoundServiceProps> = (props: AddCompoundServiceProps) => {
```
At the moment the compound service will crash the app on Sokol and xDAI networks, it would be awesome if you could disabled the service when connected to these networks 👍

Here are some details on how to setup [Sokol](https://github.com/protofire/omen-exchange/pull/1448) and [xDAI](https://github.com/protofire/omen-exchange/pull/1464) on metamask.

pRoy24 replied:
Great catch thanks I had not tested switching networks. Should be fixed now.

hexyls reviewed (changes requested):
Awesome work so far! This is a huge, complicated feature that is coming together nicely 🎉

pimato commented:
Hey @pRoy24 nice work!

Just a few UI related comments:

Remove margin-left here:
<img width="616" alt="Bildschirmfoto 2021-01-13 um 22 22 41" src="https://user-images.githubusercontent.com/5337809/104512032-e30fb680-55ed-11eb-89fa-82f6522e36e1.png">

pimato commented:
Remove margin-left here:
<img width="616" alt="Bildschirmfoto 2021-01-13 um 22 22 41" src="https://user-images.githubusercontent.com/5337809/104512092-f6228680-55ed-11eb-93be-d4e9830859ac.png">

pimato commented:
only `earn 2.64% APY` should be in #4B9E98 , currently `and earn 14.03% APY`

pimato commented:
Replace current green with: #4B9E98 
<img width="612" alt="Bildschirmfoto 2021-01-13 um 22 24 56" src="https://user-images.githubusercontent.com/5337809/104512251-384bc800-55ee-11eb-9977-5ec58d8254e1.png">

pimato commented:
`earn 0.00% APY` should have a font-weight: 500

pimato commented:
Circle button should be inactive on default!

pimato commented:
Circle icon is not 40x40, currently:
<img width="380" alt="Bildschirmfoto 2021-01-13 um 22 27 29" src="https://user-images.githubusercontent.com/5337809/104512468-8f519d00-55ee-11eb-8f8f-889fdde46c92.png">

Should:
<img width="1113" alt="Bildschirmfoto 2021-01-13 um 22 29 05" src="https://user-images.githubusercontent.com/5337809/104512612-c45def80-55ee-11eb-889b-24ea08390338.png">

pimato commented:
Circle default styling is not correct, please see: 
https://www.figma.com/file/R056tSYMjvfwaFidEBItYy/Widgets-V2?node-id=1209%3A141

pRoy24 commented:
Resolved the css issues.
Fixed the issue with the app crashing in sokol and xDai.
Made some optimizations so that the interaction is smoother.
Rebased to master.

pRoy24 commented:
> on funding page loaded, app freezes

Hey @pimato Do you mean the app stops responding or are the interactions slower?
I haven't run into this issue.

The predicted amounts change in the display will be slightly slower than normal display updates since it needs to do the additional steps of getting the exchange rate and converting the amounts from baseToken -> cToken and vice versa.
I have made some optimizations so the interaction should be smoother now, let me know if during your testing you encounter the page freezing issue again.

pimato commented:
just played around with it on Rinkeby and I am getting an error after clicking on circle button to enable it:
<img width="1045" alt="Bildschirmfoto 2021-01-15 um 16 35 19" src="https://user-images.githubusercontent.com/5337809/104746567-bd022780-574f-11eb-83da-0f1d481a0aae.png">

Steps to reproduce:

1) Connect to Rinkeby 
2) Create Market 
3) Configure Market:
<img width="613" alt="Bildschirmfoto 2021-01-15 um 16 37 04" src="https://user-images.githubusercontent.com/5337809/104746722-f20e7a00-574f-11eb-9bf4-c168e1301fea.png">
4) Click Continue
5) Click enable Compound circle button
6) BUG: App crashes

pRoy24 commented:
> just played around with it on Rinkeby and I am getting an error after clicking on circle button to enable it:
> <img alt="Bildschirmfoto 2021-01-15 um 16 35 19" width="1045" src="https://user-images.githubusercontent.com/5337809/104746567-bd022780-574f-11eb-83da-0f1d481a0aae.png">
> 
> Steps to reproduce:
> 
> 1. Connect to Rinkeby
> 2. Create Market
> 3. Configure Market:
> 
> <img alt="Bildschirmfoto 2021-01-15 um 16 37 04" width="613" src="https://user-images.githubusercontent.com/5337809/104746722-f20e7a00-574f-11eb-9bf4-c168e1301fea.png">
> 
> 4) Click Continue 5) Click enable Compound circle button 6) BUG: App crashes

You need to run Yarn install since it uses Big.js version 6.0.0 for the prec function

pimato commented:
Hey @pRoy24 , once I enable compound the pool number is jumping around:
<img width="207" alt="Bildschirmfoto 2021-01-18 um 13 13 11" src="https://user-images.githubusercontent.com/5337809/104914181-07251c00-598f-11eb-8c95-bb756bb3dc96.png">

pimato commented:
Every market info component here has a svg in 24x24. The compound svg you included is 16x16:
<img width="261" alt="Bildschirmfoto 2021-01-18 um 13 17 55" src="https://user-images.githubusercontent.com/5337809/104914705-d396c180-598f-11eb-99ad-965650ce9dce.png">

please use this svg:
[comp.svg.zip](https://github.com/protofire/omen-exchange/files/5830122/comp.svg.zip)

pimato commented:
Please remove the styling if the circle button is active, currently if you hover over it you see the border is popping up:
<img width="207" alt="Bildschirmfoto 2021-01-18 um 13 13 11" src="https://user-images.githubusercontent.com/5337809/104915185-91ba4b00-5990-11eb-9858-672d0e28aa3e.png">

pimato commented:
If you hover over the compound info the icon is turning blue, which is not according my design specification. Currently:
<img width="512" alt="Bildschirmfoto 2021-01-18 um 13 21 37" src="https://user-images.githubusercontent.com/5337809/104915279-b44c6400-5990-11eb-9323-3593254ec6d0.png">

Should:
<img width="955" alt="Bildschirmfoto 2021-01-18 um 13 23 39" src="https://user-images.githubusercontent.com/5337809/104915300-ba424500-5990-11eb-9a39-5545cba7b24a.png">

onHover color: #00897B

pimato commented:
Just looking at the price of the outcomes here:
<img width="617" alt="Bildschirmfoto 2021-01-18 um 13 27 24" src="https://user-images.githubusercontent.com/5337809/104915453-f4abe200-5990-11eb-91a7-8abb4e37c437.png">

I believe those prices are in cDAI and not in DAI right?

pimato commented:
Looks like after I done a swap, the form is not resetting which is our default.

Doing a swap:
<img width="613" alt="Bildschirmfoto 2021-01-18 um 13 31 05" src="https://user-images.githubusercontent.com/5337809/104915947-a9460380-5991-11eb-9832-4b8df48858a7.png">

after the swap is done, the inputfield is still filled with the number 2 and I see order details:
<img width="611" alt="Bildschirmfoto 2021-01-18 um 13 31 54" src="https://user-images.githubusercontent.com/5337809/104915994-b9f67980-5991-11eb-8a8c-a3009e2f6066.png">

It should reset the order form as we do on default.

pimato commented:
> Please remove the styling if the circle button is active, currently if you hover over it you see the border is popping up:
> <img alt="Bildschirmfoto 2021-01-18 um 13 13 11" width="207" src="https://user-images.githubusercontent.com/5337809/104915185-91ba4b00-5990-11eb-9858-672d0e28aa3e.png">

it looks like now the the border got removed when holding on-click:
<img width="658" alt="Bildschirmfoto 2021-01-19 um 09 27 24" src="https://user-images.githubusercontent.com/5337809/105007778-b23de000-5a38-11eb-83f3-f83f25ab35bd.png">

pimato commented:
The compound icon is still not correct, currently:
<img width="446" alt="Bildschirmfoto 2021-01-19 um 10 14 33" src="https://user-images.githubusercontent.com/5337809/105013085-2aa79f80-5a3f-11eb-86b8-a5120214d261.png">

should:
<img width="936" alt="Bildschirmfoto 2021-01-19 um 10 13 40" src="https://user-images.githubusercontent.com/5337809/105013113-3004ea00-5a3f-11eb-91e6-77f269d011ca.png">

pimato commented:
> If you hover over the compound info the icon is turning blue, which is not according my design specification. Currently:
> <img alt="Bildschirmfoto 2021-01-18 um 13 21 37" width="512" src="https://user-images.githubusercontent.com/5337809/104915279-b44c6400-5990-11eb-9323-3593254ec6d0.png">
> 
> Should:
> <img alt="Bildschirmfoto 2021-01-18 um 13 23 39" width="955" src="https://user-images.githubusercontent.com/5337809/104915300-ba424500-5990-11eb-9a39-5545cba7b24a.png">
> 
> onHover color: #00897B

you changed the normal color to #00897B, but you actually should use #00897B as the hover color for the icon and the text.

pimato reviewed (changes requested):
see comments.

Mi-Lan commented on app/src/components/market/common/add_compound_service/index.tsx:
```diff
@@ -0,0 +1,163 @@
+  -webkit-box-align: center;
+`
+
+const ServiceTextWrapper = styled.div`
+  width: 90%;
+`
+
+const ServiceCheckWrapper = styled.div`
+  width: 10%;
+  color: transparent;
```
I dont think defining static values to width here is neccesary. I would suggest just adding 
justify-content: space-between; to the parent element that is 100% and omitting width for these two values. The reason for this is because 10% is slightly more than 40px and creates this unnecessary margin...
![Screenshot 2021-01-20 at 23 28 34](https://user-images.githubusercontent.com/33226956/105248740-3ac29a80-5b77-11eb-9d28-03368aca25f0.png)

pRoy24 replied:
Fixed

Mi-Lan commented on app/src/util/networks.ts:
```diff
@@ -260,13 +260,66 @@ export const getInfuraUrl = (networkId: number): string => {
 export const knownTokens: { [name in KnownToken]: KnownTokenData } = {
   cdai: {
     symbol: 'cDAI',
-    decimals: 18,
```
![Screenshot 2021-01-20 at 23 31 04](https://user-images.githubusercontent.com/33226956/105249000-a9075d00-5b77-11eb-8564-7b272c9ce663.png)
Would be good to have consistent letter formatting for both currencies. To make them both capital or with first letter Uppercase (which is according to design...

pRoy24 replied:
Yup, makes sense refactored to include the change.

Mi-Lan reviewed (changes requested):
Small changes...
